pkg/reconciler/tenancy/initialization/apibinder_initializer_reconcile.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	// WorkspaceAPIBindingsInitialized represents the status of the initial APIBindings for the workspace.
+	WorkspaceAPIBindingsInitialized conditionsv1alpha1.ConditionType = "APIBindingsInitialized"
+	// WorkspaceInitializedWaitingOnAPIBindings is a reason for the APIBindingsInitialized condition that indicates
+	// at least one APIBinding is not ready.
+	WorkspaceInitializedWaitingOnAPIBindings = "WaitingOnAPIBindings"
+	// WorkspaceInitializedClusterWorkspaceTypeInvalid is a reason for the APIBindingsInitialized
+	// condition that indicates something is invalid with the ClusterWorkspaceType (e.g. a cycle trying
+	// to resolve all the transitive types).
+	WorkspaceInitializedClusterWorkspaceTypeInvalid = "ClusterWorkspaceTypeInvalid"
+	// WorkspaceInitializedAPIBindingErrors is a reason for the APIBindingsInitialized condition that indicates there
+	// were errors trying to initialize APIBindings for the workspace.
+	WorkspaceInitializedAPIBindingErrors = "APIBindingErrors"
+)
+
+func (b *APIBinder) reconcile(ctx context.Context, clusterWorkspace *tenancyv1alpha1.ClusterWorkspace) error {
+	logger := klog.FromContext(ctx).WithValues(
+		"clusterWorkspaceType.path", clusterWorkspace.Spec.Type.Path,
+		"clusterWorkspaceType.name", clusterWorkspace.Spec.Type.Name,
+	)
+
+	var errors []error
+	clusterName := logicalcluster.From(clusterWorkspace).Join(clusterWorkspace.Name)
+	logger.V(2).Info("initializing APIBindings for workspace")
+
+	// Start with the ClusterWorkspaceType specified by the ClusterWorkspace
+	leafCWT, err := b.getClusterWorkspaceType(logicalcluster.New(clusterWorkspace.Spec.Type.Path), string(clusterWorkspace.Spec.Type.Name))
+	if err != nil {
+		logger.Error(err, "error getting ClusterWorkspaceType")
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			WorkspaceAPIBindingsInitialized,
+			WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error getting ClusterWorkspaceType %s|%s: %v",
+			clusterWorkspace.Spec.Type.Path, clusterWorkspace.Spec.Type.Name,
+			err,
+		)
+
+		return nil
+	}
+
+	// Get all the transitive ClusterWorkspaceTypes
+	cwts, err := b.transitiveTypeResolver.Resolve(leafCWT)
+	if err != nil {
+		logger.Error(err, "error resolving transitive types")
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			WorkspaceAPIBindingsInitialized,
+			WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error resolving transitive set of cluster workspace types: %v",
+			err,
+		)
+
+		return nil
+	}
+
+	// Get current bindings
+	bindings, err := b.listAPIBindings(clusterName)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	// This keeps track of which APIBindings have been created for which APIExports
+	exportToBinding := map[apisv1alpha1.WorkspaceExportReference]*apisv1alpha1.APIBinding{}
+
+	for i := range bindings {
+		binding := bindings[i]
```

Comment:
If you're dereferencing when adding to the tracker, why do you need this value-copy?

Reply:
When you say dereferencing, are you referring to `*binding.Spec.Reference.Workspace` below? The reason I'm doing this here is for the RHS - need to store binding in the map, and not have it be the address of the range var (if I did `for _, binding := range bindings`). Right?

Reply:
Yes, but since you're not starting a goroutine etc, the `bindings[i]` construct will not hold on to the `i` - so it's not a deferred resolution, and unless I'm mistaken, you're fine.

Reply:
agreed to what @stevekuznetsov said 👍 i stumbled upon `binding := bindings[i]` as well. as no goroutine is started exportToBinding[*binding.Spec.Reference.Workspace] = binding[i]` should do the trick and is less confusing to the reader.

Reply:
in other terms: the map assignment is deterministic here :-)

Reply:
The alternative I was avoiding when choosing to do `binding := bindings[i]` is this:
```go
for _, binding := range bindings {}
```
You can't do that and store in a map. It's not really about goroutine-vs-not, it's about using references to the range variable. This (`foo := foos[i]`) is a well established pattern when you're trying to avoid the range variable issue. I also have
```go
if binding.Spec.Reference.Workspace == nil {
	continue
}
```
but if you feel strongly that we should also use bindings[i] here, we can.

Reply:
all good, yes `for _, binding := range bindings {}` is dangerous as `binding` is scoped for the whole `for` block 👍